### DISCUSSION TITLE
posts/Go/go-time/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#316] [Announcements]
https://www.liwenzhou.com/posts/Go/go-time/ 

 李文周的 Blog 提供 Go 语言中文教程,李文周博客,Go 语言学习之路是我整理的一个免费教程和视频教程，也是Go语言的入门教程，非常适合新手入门学习go语言。本文介绍了time包对时区、时间格式化、解析时间等操作。

**Comment by tangqihan** [comment]
func main() {
	now := time.Now()
	fmt.Println(now.Format("2006/01/02 15:04:05"))
	t := time.Now()
	for i := 0; i < 10; i++ {
		fmt.Println(i)
		u := time.Now()
		w := u.Sub(t)
		fmt.Println(w * 1000)
	}
}

**Comment by Neon-Rainbow** [comment]
在`go1.23`之前

```go
func After(d Duration) <-chan Time
func Tick(d Duration) <-chan Time
```

这两个函数不会被GC, 因此会导致内存泄漏.在`go1.23`之前的解决方式如下:

如果需要使用`time.After`的功能, 可以使用一个`context`, 或者`time.NewTimer`来解决,这两种方式的代码分别如下:

```go
func main() {
	// 创建一个带超时的 context
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel() // 确保资源释放

	select {
	case <-ctx.Done():
		fmt.Println("Time's up!")
	}
}
```

```go
func main() {
	// 创建一个定时器
	timer := time.NewTimer(5 * time.Second)
	defer timer.Stop() // 确保资源释放

	select {
	case <-timer.C:
		fmt.Println("Time's up!")
	}
}
```

但是`go1.23`修复了该问题,我们可以使用更加简单的方式来实现该功能

```go
func main() {
	select {
	case <-time.After(5 * time.Second): // 1.23之后不会导致泄漏
		fmt.Println("Time's up!")
	}
}
```



`time.Tick`由于内存泄漏,`go1.23`之前是使用如下的方式来解决的

```go
func main() {
	// 创建一个 Ticker，每秒触发一次
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop() // 确保退出时释放资源

	go func() {
		for t := range ticker.C {
			fmt.Println("Tick at", t)
		}
	}()

	time.Sleep(5 * time.Second)
	fmt.Println("Exiting")
}
```

而在`go1.23`我们可以使用更加简洁的方式

```go
func main() {
	select {
	case <-time.Tick(1 * time.Second):
		fmt.Println("Tick")
	}
}

```

**Reply by iwangzhy** [reply]
说明一下，这里的 time.After 、 time.Tick 两个函数不会被 GC，指的是在底层计数器触发前不会被垃圾回收器回收。
> The underlying Timer is not recovered by the garbage collector until the timer fires.
> https://pkg.go.dev/time@go1.22.12#After

在go 1.23 之后，垃圾回收器已能回收未被引用且未停止的计时器。 
> As of Go 1.23, the garbage collector can recover unreferenced, unstopped timers. There is no reason to prefer NewTimer when After will do.
> https://pkg.go.dev/time#After
